cpp/include/cudf/ast/detail/expression_evaluator.cuh

```diff
@@ -478,20 +556,33 @@ struct expression_evaluator {
           plan.data_references[plan.operator_source_indices[operator_source_index++]];
         auto const& output =
           plan.data_references[plan.operator_source_indices[operator_source_index++]];
-        type_dispatcher(lhs.data_type,
-                        detail::single_dispatch_binary_operator{},
-                        *this,
-                        output_object,
-                        left_row_index,
-                        right_row_index,
-                        lhs,
-                        rhs,
-                        output,
-                        output_row_index,
-                        op,
-                        thread_intermediate_storage);
-      } else {
-        CUDF_UNREACHABLE("Invalid operator arity.");
+        if constexpr (has_complex_type) {
+          type_dispatcher(lhs.data_type,
+                          detail::single_dispatch_binary_operator{},
+                          *this,
+                          output_object,
+                          left_row_index,
+                          right_row_index,
+                          lhs,
+                          rhs,
+                          output,
+                          output_row_index,
+                          op,
+                          thread_intermediate_storage);
+        } else {
+          type_dispatcher<id_to_primitive_type>(lhs.data_type,
+                                                detail::single_dispatch_binary_operator{},
+                                                *this,
+                                                output_object,
+                                                left_row_index,
+                                                right_row_index,
+                                                lhs,
+                                                rhs,
+                                                output,
+                                                output_row_index,
+                                                op,
+                                                thread_intermediate_storage);
+        }
```

Comment:
Similarly, just a single call to `type_dispatcher` would cover all types, since the parameters to `type_dispatcher` are identical in both calls.